art/attacks/evasion/shapeshifter.py

```diff
@@ -0,0 +1,1122 @@
+
+        # Add decay to feed_dict
+        if self.optimizer == 'RMSPropOptimizer':
+            feed_dict['decay:0'] = self.decay
+
+        # Add mask to feed_dict
+        if self.texture_as_input:
+            feed_dict['mask_input:0'] = mask
+
+        # Training loop for attack optimization
+        for _ in range(self.max_iter):
+            # Accumulate gradients
+            for _ in range(self.random_size):
+                if self.texture_as_input:
+                    # Random transformation
+                    background, image_frame, y_ = self.random_transform(x)
+
+                    # Add more to feed_dict
+                    feed_dict['background_phd:0'] = background
+                    feed_dict['image_frame_phd:0'] = image_frame
+
+                    for i in range(x.shape[0]):
+                        feed_dict['groundtruth_boxes_{}:0'.format(i)] = y_['groundtruth_boxes_list'][i]
+                        feed_dict['groundtruth_classes_{}:0'.format(i)] = y_['groundtruth_classes_list'][i]
+                        feed_dict['groundtruth_weights_{}:0'.format(i)] = y_['groundtruth_weights_list'][i]
+
+                else:
+                    # Random transformation
+                    random_transformation = self.random_transform(x)
+
+                    # Add more to feed_dict
+                    feed_dict['random_transformation_phd:0'] = random_transformation
+
+                    for i in range(x.shape[0]):
+                        feed_dict['groundtruth_boxes_{}:0'.format(i)] = y['groundtruth_boxes_list'][i]
+                        feed_dict['groundtruth_classes_{}:0'.format(i)] = y['groundtruth_classes_list'][i]
+                        feed_dict['groundtruth_weights_{}:0'.format(i)] = y['groundtruth_weights_list'][i]
+
+                # Accumulate gradients
+                self.estimator.sess.run(current_image_assign_to_input_image_op, feed_dict)
+                self.estimator.sess.run(accumulated_gradients_op, feed_dict)
+
+            # # Update feed_dict with true labels
+            # if self.texture_as_input:
+            #     for i in range(x.shape[0]):
+            #         feed_dict['groundtruth_boxes_{}:0'.format(i)] = y['groundtruth_boxes_list'][i]
+            #         feed_dict['groundtruth_classes_{}:0'.format(i)] = y['groundtruth_classes_list'][i]
+            #         feed_dict['groundtruth_weights_{}:0'.format(i)] = y['groundtruth_weights_list'][i]
+
+            # Update image/texture variable
+            self.estimator.sess.run(final_attack_optimization_op, feed_dict)
+
+            if self.texture_as_input:
+                self.estimator.sess.run(project_texture_op, feed_dict)
+
+        result = self.estimator.sess.run([current_variable, current_value], feed_dict)
+
+        return result
+
+    def _build_graph(
+        self,
+        initial_shape: Tuple[int, ...],
+        custom_loss: Optional["Tensor"] = None,
+        rendering_function: Optional["Callable"] = None
```

Comment:
I think `rendering_function` cannot be None. Maybe we should remove the default here?

Reply:
It can be None when the option texture_as_input is False.